File: Algorithm_problem_solving/Baek-joon/1600/1600.py
import sys
from queue import Queue

input = sys.stdin.readline

jump = ((2, -1), (2, 1), (1, 2), (-1, 2), (-2, 1), (-2, -1), (1, -2), (-1, -2))
way = ((1, 0), (-1, 0), (0, 1), (0, -1))


def getMinMove(k, w, h, board):
    que = Queue();
    visited = [[float('inf')] * w for _ in range(h)]
    curr = (0, 0, 0, 0)  # y, x, 이동, 점프
    que.put(curr)
    visited[0][0] = 0

    while que:
        curr = que.get()
        if curr[0] == h - 1 and curr[1] == w - 1:
            return curr[2]
        for nextMv in way:
            nextY = curr[0] + nextMv[0]
            nextX = curr[1] + nextMv[1]
            if nextY < 0 or nextX < 0 or nextY >= h or nextX >= w:
                continue
            if board[nextY][nextX] == 1 or visited[nextY][nextX] <= curr[3]:
                continue
            que.put((nextY, nextX, curr[2] + 1, curr[3]))
            visited[nextY][nextX] = curr[3]
        if curr[3] < k:
            for nextMv in jump:
                nextY = curr[0] + nextMv[0]
                nextX = curr[1] + nextMv[1]
                if nextY < 0 or nextX < 0 or nextY >= h or nextX >= w:
                    continue
                if board[nextY][nextX] == 1 or visited[nextY][nextX] <= curr[3] + 1:
                    continue
                que.put((nextY, nextX, curr[2] + 1, curr[3] + 1))
                visited[nextY][nextX] = curr[3] + 1


if __name__ == "__main__":
    k = int(input())
    w, h = map(int, input().strip().split())
    board = [[0]*w for _ in range(h)]

    for i in range(h):
        inline = list(map(int, input().strip().split()))
        for j in range(w) :
            board[i][j] = inline[j]
    print(getMinMove(k, w, h, board))

File: Algorithm_problem_solving/Baek-joon/1600/1600.kt
import java.io.IOException

// https://www.acmicpc.net/problem/1600
fun main() {
    val availableJump: Int = readlnOrNull()?.toIntOrNull() ?: throw IOException("에러 발생")
    val (width, height) = readLine()!!.trim().split(" ").map { it.toInt() }

    val matrix = Array(height) { Array(width) { 0 } }

    for (i in 0 until height) {
        val tempInput = readLine()!!.trim().split(" ").map { it.toInt() }.toTypedArray()
        for ((j, value) in tempInput.withIndex()) { //indices -> 0..2
            matrix[i][j] = value
        }
    }

    println(MonkeyJump(availableJump, matrix).getShortestJumping())
}

class MonkeyJump(
    val k: Int, val matrix: Array<Array<Int>>
) {
    val w: Int = matrix[0].size
    val h: Int = matrix.size
    var walk = arrayOf(
        intArrayOf(1, 0), intArrayOf(-1, 0), intArrayOf(0, 1), intArrayOf(0, -1)
    )
    val jump: Array<Array<Int>> = arrayOf(
        arrayOf(2, -1),
        arrayOf(2, 1),
        arrayOf(1, 2),
        arrayOf(-1, 2),
        arrayOf(-2, 1),
        arrayOf(-2, -1),
        arrayOf(1, -2),
        arrayOf(-1, -2)
    )

    class MoveInfo(
        val y: Int, val x: Int, val jumpCount: Int, val moveCount: Int
    )

    fun getShortestJumping(): Int {
        val visited = Array(h) { Array(w) { -1 } }
        val queue = ArrayDeque<MoveInfo>()

        queue.add(MoveInfo(0, 0, 0, 0))
        visited[0][0] = 0

        while (queue.isNotEmpty()) {
            val curr = queue.removeFirst()
            if (curr.y == h - 1 && curr.x == w - 1) {
                return curr.moveCount
            }
            for ((moveY, moveX) in walk) {
                val nextY = curr.y + moveY
                val nextX = curr.x + moveX
                if (nextY < 0 || nextY >= h || nextX < 0 || nextX >= w) continue
                if (matrix[nextY][nextX] == 1) continue
                if (visited[nextY][nextX] >= curr.jumpCount) continue
                queue.add(MoveInfo(nextY, nextX, curr.jumpCount, curr.moveCount + 1))
                visited[nextY][nextX] = curr.jumpCount
            }
            if (curr.jumpCount < k) {
                for ((moveY, moveX) in jump) {
                    val nextY = curr.y + moveY
                    val nextX = curr.x + moveX
                    if (nextY < 0 || nextY >= h || nextX < 0 || nextX >= w) continue
                    if (matrix[nextY][nextX] == 1) continue
                    if (visited[nextY][nextX] >= curr.jumpCount) continue
                    queue.add(MoveInfo(nextY, nextX, curr.jumpCount + 1, curr.moveCount + 1))
                    visited[nextY][nextX] = curr.jumpCount + 1
                }
            }
        }

        return -1
    }
}